### ISSUE TITLE
[java] SingularField false-positive with field used concurrently

### ISSUE THREAD

**Issue by kohlschuetter** [pmd/pmd#4616]
**Affects PMD Version:** 
7.0.0-SNAPSHOT

**Rule:**
[https://pmd.github.io/pmd/pmd_rules_java_design.html#singularfield](SingularField)

**Description:**

**Code Sample demonstrating the issue:**

```java
// AFTIPCTopologyWatcher.java
public class AFTIPCTopologyWatcher implements Closeable {
  private final int defaultTimeout;
  private final AFTIPCDatagramChannel channel;
  private final Selector selector;
  private boolean doLoop;
  private final AtomicBoolean running = new AtomicBoolean(false);


  public final void runLoop() throws IOException {
    if (!running.compareAndSet(false, true)) {
      throw new IllegalStateException("Already running");
    }

    ByteBuffer buf = ByteBuffer.allocate(64);
    SelectionKey key = channel.register(selector, SelectionKey.OP_READ);

    try {
      doLoop = true;
      while (!Thread.interrupted() && doLoop) {
        int n = selector.select();
        if (!key.isValid() || !doLoop) {
          break;
        }
        if (n > 0) {
          channel.receive(buf);
          @SuppressWarnings("cast")
          AFTIPCTopologyEvent event = AFTIPCTopologyEvent.readFromBuffer((ByteBuffer) buf.flip());
          onEvent(event);
          buf.clear();
        }
      }
    } finally {
      key.cancel();
      running.set(false);
    }
  }

  protected void onEvent(AFTIPCTopologyEvent event) throws IOException {
  }


  public final void stopLoop() {
    doLoop = false;
    selector.wakeup();
  }

}
```

**Expected outcome:**

PMD reports a violation, but that's wrong. That's a false positive.
```
AFTIPCTopologyWatcher.java:39:	SingularField:	Perhaps 'doLoop' could be replaced by a local variable.
```

`doLoop` is checked/set in two different methods, so it cannot be replaced by a local variable.

The error does not occur if we change `doLoop` to a `final AtomicBoolean`

**Running PMD through:** CLI

File and ruleset attached.
[file-and-ruleset1.zip](https://github.com/pmd/pmd/files/11909516/file-and-ruleset1.zip)

**Comment by oowekyala** [comment]
Same preface as https://github.com/pmd/pmd/issues/4617#issuecomment-1615979261, the rule doesn't understand threading. I don't think we can make the rule ignore this case without also ignoring many legitimate cases. Maybe we can ignore volatile fields, but this one is not volatile anyway. Suppression is probably the way to go here.

**Comment by kohlschuetter** [comment]
Can we perhaps invent and add an annotation like `@ConcurrentUse` or something that indicates that `doLoop` is used from a multithreaded-context, and then suppress that error?

I'd argue myself that `AtomicBoolean` is a better solution, since it guarantees early-termination (unlike leaving it as-is or even using `volatile`), but that's a different problem.

**Comment by oowekyala** [comment]
> Can we perhaps invent and add an annotation like @ConcurrentUse or something that indicates that doLoop is used from a multithreaded-context, and then suppress that error?

You can define your own annotation and make the rule recognize it by setting the property `ignoredAnnotations`. However that annotation should be codebase specific, PMD has no business recommending non-standard annotations be used to suppress warnings. You can achieve the same effect with `@SuppressWarnings` anyway.

> I'd argue myself that AtomicBoolean is a better solution, since it guarantees early-termination (unlike leaving it as-is or even using volatile), but that's a different problem.

Yes, using AtomicBoolean may be the way to go for you for this specific warning. Since you reported this would fix the warning, I will close this as "won't fix". I recommend [suppressing the warning](https://pmd.github.io/pmd/pmd_userdocs_suppressing_warnings.html#annotations) or using `ignoredAnnotations` to get around this in the future. I will update the doc of the rule to point that out.
